client/internal/callbacks.go

```diff
@@ -0,0 +1,125 @@
+package internal
+
+import (
+	"context"
+
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+type CallbacksStruct struct {
```

Comment:
I'm not opposed to this implementation. Usage of this struct is more intuitive to me than accessing functions in a map (e.g. `w.settings.Callbacks.OnConnectFailed(err)`).

Reply:
> This struct feels a little overkill and could potentially, to me this looks like a map of methods that can have values set?

> If that is the case, we can simplify with an enum and a map :)

I disagree. That would make impossible to make functions correctly strongly typed. With this approach every callback function has the appropriate signature with the right parameter  and return types.

Reply:
I think the the things that erks me is the verbosity since this isn't typically I had seen in golang.

I am not sure if this is going to be an object that is passed around or scoped within a stack.

I hadn't checked to see how it is being used but I imagine you could do something like this:

```golang
type Callback func(ctx context.Context) error

type Callbacks []Callback

func OnConnectionResolve(err error) Callback {
    return func(ctx context.Context) error {
         // Do something with error and attempt connection
        return nil
    }
}
```

I can see this struct getting infinitely big with arbitrary function signatures.

Reply:
I do not see a good, type-safe way to do what you suggest. Perhaps there is a way but I fail to see it. I'll go ahead with the current approach for now to make progress on the implementation but if you can suggest an improvement I will be happy to review it if you submit a PR.